### PR TITLE
feat(compiler)!: explicit lift qualification statement

### LINKED ISSUE #6080: The syntax around `lift()` statements is confusing
### Use Case

This issue is for tracking the implementation for https://github.com/winglang/wing/pull/5951

### Proposed Solution

_No response_

### Implementation Notes

_No response_

### Component

Language Design

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

### PR DESCRIPTION
**Pull request by yoav-steinberg**
fixes #6080

Initial implementation of the explicit lift statement as defined in the [rfc](https://github.com/winglang/wing/blob/main/docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md).

```wing
let bucket = new cloud.Bucket();
inflight () => {
  let b = bucket; // Assing preflight object to inflight variable
  lift {bucket: put} {
    b.put("k","v"); // Within this block we can use the inflight variable because of the explicit lifting above
  }
  b.put("k","v"); // This is an error, because it's outside the explicit lift block and we can't figure out who `b` is
  
  // We can also define multiple lifts in a block and multiple ops per lift
  lift {bucket1: [put, get], bucket2: delete} { ... }
  
  // We can nest lift blocks
  lift {bucket1: put} {
    ..
    lift {bucket2: get} {
      ..
    }
  }

  // Type checker validates passed methods are really part of the inflight interface of the object
  lift {bucket: leak} {}
              //^ err: leak isn't an inflight method on type Bucket
}
```

Breaking change: this is instead of the `lift()` builtin macro that was used for the same purpose. 

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingc/src/parser.rs [hunk 1]
```diff
@@ -637,6 +638,43 @@ impl<'s> Parser<'s> {
 		})
 	}
 
+	fn build_lift_statement(&self, statement_node: &Node, phase: Phase) -> DiagnosticResult<StmtKind> {
+		// Lift statements are only legal in inflight
+		if phase != Phase::Inflight {
+			return self.with_error("Lift statements are only allowed in inflight phase", statement_node);
+		}
```
Comment (MarkMcCulloh) [inline libs/wingc/src/parser.rs:645]: Is there a technical reason for this restriction? Could/Should this work?

```wing
// I could only think of a contrived example
let b = new cloud.Bucket();
lift {b: [get]} {
  let f = new cloud.Function(inflight () => {
    b.get("blah");
  })
}
```

Comment (yoav-steinberg) [reply libs/wingc/src/parser.rs:645]: Interesting. Today the lift mechanism stores lift information per method (in your example the implicit `handle` method). So when specifying `lift` or when actually doing something like `preflight_bucket.put()` the lift visitor knows in which method we're in and stores the relevant information for that method.

What you're suggesting implies that you can specify explicit lifts that are assigned to multiple methods:
```wing
lift {bucket: get} {
  class Foo {
    inflight foo_m() {
      let b = bucket;
      b.get("hey");
    }
  }
  class Boo {
    inflight boo_m() {
      let b = bucket;
      b.get("hey");
    }
  }
```

I'll need to modify the lift visitor to get that working..
I suggest a future improvement, if at all.

File: libs/wingc/src/type_check.rs [hunk 2]
```diff
@@ -6288,6 +6253,76 @@ impl<'a> TypeChecker<'a> {
...
+				self.spanned_error(
+					&qual.obj,
+					format!("Expected a preflight object, but found {obj_phase} expression instead"),
+				);
+			}
+			// Make sure the object type is a preflight type
+			if !obj_type.is_preflight_object_type() {
+				self.spanned_error(
+					&qual.obj,
+					format!("Expected a preflight object type, but found {obj_type} instead"),
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:6281]: (nit) I feel like "preflight object type" could be confusing. Maybe we can mention something in the diagnostic as a hint, or reword this in some way

File: tools/hangar/__snapshots__/invalid.ts.snap [hunk 3]
```diff
@@ -3847,25 +3879,39 @@ Duration <DURATION>"
...
-   --> ../../../examples/tests/invalid/simulator_permissions.test.w:20:8
+error: Unknown parser error
+   --> ../../../examples/tests/invalid/simulator_permissions.test.w:20:3
    |
 20 |   lift(buckets.at(0), [\\"list\\"]);
-   |        ^^^^^^^^^^^^^ Expected a preflight object as first argument to \`lift\` builtin, found inflight expression instead
+   |   ^^^^ Unknown parser error
+
+
+error: Expression of type \\"Bucket\\" references an unknown preflight object, can't qualify its capabilities. Use \`lift()\` to explicitly qualify the preflight object to disable this error.
```
Comment (Chriscbr) [inline tools/hangar/__snapshots__/invalid.ts.snap:3910]: Can we add a hint to this diagnostic that links to the language reference section? I imagine here is a good place: https://www.winglang.io/docs/concepts/inflights#explicit-lift-qualification

Comment (Chriscbr) [inline tools/hangar/__snapshots__/invalid.ts.snap:3910]:
```suggestion
error: Expression of type \\"Bucket\\" references an unknown preflight object, can't qualify its capabilities. Use a \`lift\` block to explicitly qualify the preflight object and disable this error.
```

File: libs/wingc/src/type_check.rs [hunk 4]
```diff
@@ -6288,6 +6253,76 @@ impl<'a> TypeChecker<'a> {
...
+							self.spanned_error(
+								op,
+								"Only instance (non-static) members may qualify the lift".to_string(),
+							);
+						}
+					}
+					Some(_) => panic!("expected object envs to only have variables"),
+					None => self.spanned_error_with_annotations(
+						op,
+						format!("Object of type {obj_type} does not have an inflight member named {op}"),
```
Comment (Chriscbr) [inline libs/wingc/src/type_check.rs:6306]:
```suggestion
						format!("Object of type {obj_type} does not have an inflight member named \"{op}\""),
```

File: libs/wingc/src/parser.rs [hunk 5]
```diff
@@ -637,6 +638,37 @@ impl<'s> Parser<'s> {
 		})
 	}
 
+	fn build_lift_statement(&self, statement_node: &Node, phase: Phase) -> DiagnosticResult<StmtKind> {
+		// Lift statements are only legal in inflight
+		if phase != Phase::Inflight {
+			return self.with_error("Lift statements are only allowed in inflight phase", statement_node);
```
Comment (Chriscbr) [inline libs/wingc/src/parser.rs:644]:
```suggestion
			return self.with_error("Lift blocks are only allowed in inflight phase", statement_node);
```

**Pull request merged**